### ISSUE TITLE
NEXT_PUBLIC_ variables are undefined on production build

### ISSUE THREAD

**Issue by vsDizzy** [vercel/next.js#25774]
### What version of Next.js are you using?

latest

### What version of Node.js are you using?

latest

### What browser are you using?

latest

### What operating system are you using?

docker

### How are you deploying your application?

via console

### Describe the Bug

NEXT_PUBLIC_ variables are undefined on production builds.

### Expected Behavior

NEXT_PUBLIC_ variables are defined

### To Reproduce

clone https://github.com/vsDizzy/next-env-bug

npm run dev1
open http://localhost:3000/
F12, see `TEST: test` in console

npm run build
npm run start1
open http://localhost:3000/
F12, see `TEST: undefined` in console

**Comment by giovannibenussi** [comment]
The issue you have is because environment variables need to be set at *build time* instead of when you run the app.
In your case, you have the following commands:

- Build: `next build`
- Start: `NEXT_PUBLIC_TEST=test next start`

Instead, they need to be:

- Build: `NEXT_PUBLIC_TEST=test next build`
- Start: `next start`

This is because when you build your app with `yarn build`, it is generating static HTML + JSON ([Static Generation](https://nextjs.org/docs/basic-features/pages#two-forms-of-pre-rendering)) and all `process.env.` variables get replaced with their values.

![image](https://user-images.githubusercontent.com/1259399/120801418-1f956d80-c50f-11eb-8427-96bffc0d564f.png)

 It would be different if your app runs on a dynamic server, in which case your code should work, but it's not the case.

[This behavior is mentioned in the docs too](https://nextjs.org/docs/basic-features/environment-variables#exposing-environment-variables-to-the-browser):

> This loads process.env.NEXT_PUBLIC_ANALYTICS_ID into the Node.js environment automatically, allowing you to use it anywhere in your code. The value will be inlined into JavaScript sent to the browser because of the NEXT_PUBLIC_ prefix. This inlining occurs at build time, so your various NEXT_PUBLIC_ envs need to be set when the project is built.

**Comment by vsDizzy** [comment]
Yes, now it works. Thank you. @giovannibenussi

**Comment by eeeman1** [comment]
how can i add third or fourth value? In addition to development or production, I also use staging

**Comment by giovannibenussi** [comment]
This is outside the scope of Next.js. However, you can create a `.env.staging` file and rename it to `.env.local` in your continuous integration service. Let me know if you need more help!

**Comment by eeeman1** [comment]
> This is outside the scope of Next.js. However, you can create a `.env.staging` file and rename it to `.env.local` in your continuous integration service.
> 
> If you use Vercel, you can [set up a custom staging environment](https://vercel.com/support/articles/set-up-a-staging-environment-on-vercel).
> 
> Let me know if you need more help!

help is still needed, if I create a .env.local file it will work for both staging and production

**Comment by eeeman1** [comment]
`   
 "dev": "PORT=4111 next", // <- .env.development
    "build": "next build",
    "staging": "next start", // <- ??????
    "production":  next start" // <- .env.production
`

**Comment by giovannibenussi** [comment]
Checking their [source code](https://github.com/vercel/next.js/blob/1442b582f59875185f96a3d3b9d615ee7d88d721/packages/next-env/index.ts#L81), you can only have only one of `.env.development` or `.env.production` plus a `.env.local` file.

As a workaround, you could have a `.env.development` and a `.env.staging` file and update your scripts so it copies the staging environment file to `.env.local` before starting the server and clean it up before start every other environment like this:

```
"dev": "echo '' > .env.local && PORT=4111 next", // <- .loads only env.development
"build": "echo '' > .env.local && next build",
"staging": "cp .env.staging .env.local && next start", // <- loads only .env.staging
"production": echo '' > .env.local && next start" // <- loads only .env.production
```

**Comment by eeeman1** [comment]
thanks, it looks like this is what you need 👍

**Comment by balazsorban44** [comment]
This issue has been automatically locked due to no recent activity. If you are running into a similar issue, please create a new issue with the steps to reproduce. Thank you.
